File: samples/Tutorial/Movies.WPF/MainWindow.Presenter.cs
using DevZest.Data;
using DevZest.Data.Presenters;
using DevZest.Data.Views;
using System.Threading;
using System.Threading.Tasks;
using System.Windows.Controls;

namespace Movies.WPF
{
    partial class MainWindow
    {
        public static class Styles
        {
            public static readonly StyleId CheckBox = new StyleId(typeof(MainWindow));
            public static readonly StyleId LeftAlignedTextBlock = new StyleId(typeof(MainWindow));
            public static readonly StyleId RightAlignedTextBlock = new StyleId(typeof(MainWindow));
        }

        public sealed class Presenter : DataPresenter<Movie>
        {
            public interface IFilter
            {
                string Text { get; set; }
            }

            public Presenter(IFilter filter)
            {
                _filter = filter;
            }

            private readonly IFilter _filter;

            private Task<DataSet<Movie>> LoadDataAsync(CancellationToken ct)
            {
                return App.ExecuteAsync(db => db.GetMoviesAsync(_filter.Text, ct));
            }

            public void ShowAsync(DataView dataView)
            {
                ShowAsync(dataView, LoadDataAsync);
            }

            public Task RefreshAsync(bool clearFiler)
            {
                if (clearFiler)
                    _filter.Text = null;
                return RefreshAsync(LoadDataAsync);
            }

            protected override void BuildTemplate(TemplateBuilder builder)
            {
                builder.GridColumns("20", "*", "Auto", "Auto", "Auto")
                .GridRows("Auto", "Auto")
                .Layout(Orientation.Vertical)
                .WithFrozenTop(1)
                .AddBinding(0, 0, this.BindToCheckBox().WithStyle(Styles.CheckBox))
                .AddBinding(1, 0, _.Title.BindToColumnHeader())
                .AddBinding(2, 0, _.ReleaseDate.BindToColumnHeader())
                .AddBinding(3, 0, _.Genre.BindToColumnHeader())
                .AddBinding(4, 0, _.Price.BindToColumnHeader())
                .AddBinding(0, 1, _.BindToCheckBox().WithStyle(Styles.CheckBox))
                .AddBinding(1, 1, _.Title.BindToTextBlockHyperlink(Commands.Open).WithStyle(Styles.LeftAlignedTextBlock))
                .AddBinding(2, 1, _.ReleaseDate.BindToTextBlock("{0:d}").WithStyle(Styles.RightAlignedTextBlock))
                .AddBinding(3, 1, _.Genre.BindToTextBlock().WithStyle(Styles.LeftAlignedTextBlock))
                .AddBinding(4, 1, _.Price.BindToTextBlock("{0:C}").WithStyle(Styles.RightAlignedTextBlock));
            }
        }
    }
}

File: samples.vb/Tutorial/Movies.WPF/MainWindow.Presenter.vb
Imports System.Threading
Imports DevZest.Data
Imports DevZest.Data.Presenters
Imports DevZest.Data.Views

Partial Class MainWindow
    Public NotInheritable Class Styles
        Public Shared ReadOnly CheckBox As New StyleId(GetType(MainWindow))
        Public Shared ReadOnly LeftAlignedTextBlock As New StyleId(GetType(MainWindow))
        Public Shared ReadOnly RightAlignedTextBlock As New StyleId(GetType(MainWindow))
    End Class

    Public NotInheritable Class Presenter
        Inherits DataPresenter(Of Movie)

        Public Interface IFilter
            Property Text() As String
        End Interface

        Public Sub New(filter As IFilter)
            _filter = filter
        End Sub

        Private ReadOnly _filter As IFilter

        Private Function LoadDataAsync(ct As CancellationToken) As Task(Of DataSet(Of Movie))
            Return Application.ExecuteAsync(Function(db) db.GetMoviesAsync(_filter.Text, ct))
        End Function

        Public Overloads Sub ShowAsync(dataView As DataView)
            ShowAsync(dataView, AddressOf LoadDataAsync)
        End Sub

        Public Overloads Function RefreshAsync(clearFilder As Boolean) As Task
            If clearFilder Then
                _filter.Text = Nothing
            End If
            Return RefreshAsync(AddressOf LoadDataAsync)
        End Function

        Protected Overrides Sub BuildTemplate(builder As TemplateBuilder)
            builder.GridColumns("20", "*", "Auto", "Auto", "Auto") _
                .GridRows("Auto", "Auto") _
                .Layout(Orientation.Vertical) _
                .WithFrozenTop(1) _
                .AddBinding(0, 0, Me.BindToCheckBox().WithStyle(Styles.CheckBox)) _
                .AddBinding(1, 0, Entity.Title.BindToColumnHeader()) _
                .AddBinding(2, 0, Entity.ReleaseDate.BindToColumnHeader()) _
                .AddBinding(3, 0, Entity.Genre.BindToColumnHeader()) _
                .AddBinding(4, 0, Entity.Price.BindToColumnHeader()) _
                .AddBinding(0, 1, Entity.BindToCheckBox().WithStyle(Styles.CheckBox)) _
                .AddBinding(1, 1, Entity.Title.BindToHyperlink(Commands.Open).WithStyle(Styles.LeftAlignedTextBlock)) _
                .AddBinding(2, 1, Entity.ReleaseDate.BindToTextBlock("{0:d}").WithStyle(Styles.RightAlignedTextBlock)) _
                .AddBinding(3, 1, Entity.Genre.BindToTextBlock().WithStyle(Styles.LeftAlignedTextBlock)) _
                .AddBinding(4, 1, Entity.Price.BindToTextBlock("{0:C}").WithStyle(Styles.RightAlignedTextBlock))
        End Sub
    End Class
End Class
